projs/shadow/shadow-engine/core/inc/event-bus/event_bus.h

```diff
@@ -0,0 +1,139 @@
+#pragma once
+
+#include <string>
+#include <functional>
+#include <map>
+#include <memory>
+#include <any>
+
+#include "SHObject.h"
+#include "events.h"
+
+namespace ShadowEngine::EventBus {
+
+    template<class T, class U>
+    concept Derived = std::is_base_of<T, U>::value;
+
+    template<typename T>
+    concept SHObjectConcept = Derived<ShadowEngine::SHObject, T>;
+
+
+
+
+    template<typename T>
+    concept EventType = Derived<Event, T>;
+
+    template<EventType T>
+    struct Subscription {
+        using Callback = std::function<const void(T &)>;
+
+        Callback callback;
+
+        std::weak_ptr<ShadowEngine::SHObject> binding;
+
+        explicit Subscription(Callback call): callback(call){}
+    };
+
+    class EventDispatcherBase {
+    public:
+        virtual ~EventDispatcherBase() = default;
+
+        virtual void call(ShadowEngine::SHObject &obj) = 0;
+    };
+
+    template<EventType T>
+    class EventDispatcher : public EventDispatcherBase {
+    public:
+        using SubRef = std::shared_ptr<Subscription<T>>;
+
+    protected:
+        using SubscriptionList = std::vector<SubRef>;
+
+        SubscriptionList subscriptionList;
+
+    public:
+        std::shared_ptr<Subscription<T>> subscribe(const std::shared_ptr<ShadowEngine::SHObject> &binding, std::function<const void(T &)> func) {
+            auto sub = std::make_shared<Subscription<T>>(func);
+            subscriptionList.push_back(sub);
+            return sub;
+        }
+
+        void unsubscribe(const std::shared_ptr<Subscription<T>> ref) {
+            subscriptionList.erase(std::remove(subscriptionList.begin(), subscriptionList.end(), ref), subscriptionList.end());
+        }
+
+        void call(ShadowEngine::SHObject &obj) override {
+            this->call_impl((T&)obj);
```

Comment:
In the case where the implementation of call is a simple redirect, and call_impl cannot be called externally through any means... why redirect at all? only call will ever be run in this dispatcher, so move the loop below up into this.